consensus/state_processing/src/block_replayer.rs

```diff
@@ -0,0 +1,292 @@
+    fn get_state_root(
+        &mut self,
+        slot: Slot,
+        blocks: &[SignedBeaconBlock<E>],
+        i: usize,
+    ) -> Result<Option<Hash256>, Error> {
+        // If we don't care about state roots then return immediately.
+        if self.state_root_strategy == StateRootStrategy::Inconsistent {
+            return Ok(Some(Hash256::zero()));
+        }
+
+        // If a state root iterator is configured, use it to find the root.
+        if let Some(ref mut state_root_iter) = self.state_root_iter {
+            let opt_root = state_root_iter
+                .take_while(|res| res.as_ref().map_or(true, |(_, s)| *s <= slot))
+                .find(|res| res.as_ref().map_or(true, |(_, s)| *s == slot))
+                .transpose()?;
+
+            if let Some((root, _)) = opt_root {
+                return Ok(Some(root));
+            }
+        }
+
+        // Otherwise try to source a root from the previous block.
+        if i > 0 {
+            if let Some(prev_block) = blocks.get(i.safe_sub(1).map_err(BlockReplayError::from)?) {
+                if prev_block.slot() == slot {
+                    return Ok(Some(prev_block.state_root()));
+                }
+            }
+        }
+
+        self.state_root_miss = true;
+        Ok(None)
+    }
+
+    /// Apply `blocks` atop `self.state`, taking care of slot processing.
+    ///
+    /// If `target_slot` is provided then the state will be advanced through to `target_slot`
+    /// after the blocks have been applied.
+    pub fn apply_blocks(
+        mut self,
+        blocks: Vec<SignedBeaconBlock<E>>,
+        target_slot: Option<Slot>,
+    ) -> Result<Self, Error> {
+        for (i, block) in blocks.iter().enumerate() {
+            if block.slot() <= self.state.slot() {
+                continue;
+            }
+
+            while self.state.slot() < block.slot() {
+                if let Some(ref mut pre_slot_hook) = self.pre_slot_hook {
+                    pre_slot_hook(&mut self.state)?;
+                }
+
+                let state_root = self.get_state_root(self.state.slot(), &blocks, i)?;
+                per_slot_processing(&mut self.state, state_root, self.spec)
+                    .map_err(BlockReplayError::from)?;
+
+                if let Some(ref mut post_slot_hook) = self.post_slot_hook {
+                    let is_skipped_slot = self.state.slot() < block.slot();
+                    post_slot_hook(&mut self.state, is_skipped_slot)?;
+                }
+            }
+
+            if let Some(ref mut pre_block_hook) = self.pre_block_hook {
+                pre_block_hook(&mut self.state, block)?;
+            }
+
+            // Skip block root verification: this is faster and also compatible with inaccurate
+            // state root replay.
+            per_block_processing(
+                &mut self.state,
+                block,
+                None,
+                self.block_sig_strategy,
+                VerifyBlockRoot::False,
```

Comment:
Is skipping block root verification just to save cycles on the tree hashing? I'm wondering if it's worth at least checking it on the first block. It's generally without single-digit milliseconds in my experience, perhaps the guarantee on consistency is worth it?

Reply:
This is mostly for the inconsistent state root case, but I figured I'd leave it off in general for speed (1ms repeated over 2048 slots is still ~2 seconds). I like the idea of checking it for the first block though, so I'll change it to:

```rust
let verify_block_root = if i == 0 && self.state_root_strategy == StateRootStrategy::Accurate {
    VerifyBlockRoot::True
} else {
    VerifyBlockRoot::False
};
```

Reply:
Block root verification is now on by default, and one can opt in to the "check just the first" behaviour by calling `minimal_block_root_verification`